Improve diagnostic messages for range patterns.

michaelsproul opened the pull request:
Part of #24407.

Currently the diagnostics for range patterns are a bit wrong:

```rust
fn main() {
    match 5u32 {
        0 ... 10 => (),
        'a' ... 10 => (),
        10 ... 'z' => (),
        "what" ... 10 => (),
        "what" ... "well" => (),
        10 ... "what" => ()
    }
}
```

```
range.rs:4:9: 4:19 error: mismatched types in range:
 expected integral variable,
    found char [E0211]
range.rs:4         'a' ... 10 => (),
                   ^~~~~~~~~~
range.rs:4:9: 4:16 error: only char and numeric types are allowed in range [E0029]
range.rs:4         'a' ... 10 => (),
                   ^~~~~~~
range.rs:4:9: 4:19 error: mismatched types:
 expected `u32`,
    found `char`
(expected u32,
    found char) [E0308]
range.rs:4         'a' ... 10 => (),
                   ^~~~~~~~~~
range.rs:5:9: 5:19 error: mismatched types in range:
 expected char,
    found integral variable [E0211]
range.rs:5         10 ... 'z' => (),
                   ^~~~~~~~~~
range.rs:5:9: 5:15 error: only char and numeric types are allowed in range [E0029]
range.rs:5         10 ... 'z' => (),
                   ^~~~~~
range.rs:6:9: 6:22 error: mismatched types in range:
 expected integral variable,
    found &-ptr [E0211]
range.rs:6         "what" ... 10 => (),
                   ^~~~~~~~~~~~~
range.rs:6:9: 6:19 error: only char and numeric types are allowed in range [E0029]
range.rs:6         "what" ... 10 => (),
                   ^~~~~~~~~~
range.rs:6:9: 6:22 error: mismatched types:
 expected `u32`,
    found `&'static str`
(expected u32,
    found &-ptr) [E0308]
range.rs:6         "what" ... 10 => (),
                   ^~~~~~~~~~~~~
range.rs:7:9: 7:19 error: only char and numeric types are allowed in range [E0029]
range.rs:7         "what" ... "well" => (),
                   ^~~~~~~~~~
range.rs:7:9: 7:26 error: mismatched types:
 expected `u32`,
    found `&'static str`
(expected u32,
    found &-ptr) [E0308]
range.rs:7         "what" ... "well" => (),
                   ^~~~~~~~~~~~~~~~~
range.rs:8:9: 8:22 error: mismatched types in range:
 expected &-ptr,
    found integral variable [E0211]
range.rs:8         10 ... "what" => ()
                   ^~~~~~~~~~~~~
range.rs:8:9: 8:15 error: only char and numeric types are allowed in range [E0029]
range.rs:8         10 ... "what" => ()
                   ^~~~~~
error: aborting due to 12 previous errors
```

The problems here are:

1. The type of the end of the range is used to predict the type of the start (only mildly counter intuitive).
2. E0029 is erroneously generated for `char ... num` and `num ... char`.
2. `u32` is mentioned.
3. Errors which are essentially the same are reported multiple times.

I've attempted to fix this by checking the requirements in a different order. The output I've achieved for the above example is:

```
/home/michael/Temp/range.rs:4:17: 4:22 error: mismatched types in range:
 expected char,
    found integral variable [E0211]
/home/michael/Temp/range.rs:4         'a' ... 10 => (),
                                              ^~~~~
/home/michael/Temp/range.rs:5:16: 5:22 error: mismatched types in range:
 expected integral variable,
    found char [E0211]
/home/michael/Temp/range.rs:5         10 ... 'z' => (),
                                             ^~~~~~
/home/michael/Temp/range.rs:6:9: 6:19 error: only char and numeric types are allowed in range [E0029]
/home/michael/Temp/range.rs:6         "what" ... 10 => (),
                                      ^~~~~~~~~~
/home/michael/Temp/range.rs:6:9: 6:19 help: run `rustc --explain E0029` to see a detailed explanation
/home/michael/Temp/range.rs:6:9: 6:19 note: Start type: &'static str
End type: _
/home/michael/Temp/range.rs:6         "what" ... 10 => (),
                                      ^~~~~~~~~~
/home/michael/Temp/range.rs:7:9: 7:26 error: only char and numeric types are allowed in range [E0029]
/home/michael/Temp/range.rs:7         "what" ... "well" => (),
                                      ^~~~~~~~~~~~~~~~~
/home/michael/Temp/range.rs:7:9: 7:26 help: run `rustc --explain E0029` to see a detailed explanation
/home/michael/Temp/range.rs:7:9: 7:26 note: Start type: &'static str
End type: &'static str
/home/michael/Temp/range.rs:7         "what" ... "well" => (),
                                      ^~~~~~~~~~~~~~~~~
/home/michael/Temp/range.rs:8:16: 8:25 error: only char and numeric types are allowed in range [E0029]
/home/michael/Temp/range.rs:8         10 ... "what" => ()
                                             ^~~~~~~~~
/home/michael/Temp/range.rs:8:16: 8:25 help: run `rustc --explain E0029` to see a detailed explanation
/home/michael/Temp/range.rs:8:16: 8:25 note: Start type: _
End type: &'static str
/home/michael/Temp/range.rs:8         10 ... "what" => ()
                                             ^~~~~~~~~
error: aborting due to 5 previous errors
```

I think this is already tonnes better, but the `Start type/End type` stuff could be neater. I don't think there's really any need to start a `note:` block but I wanted to get some feedback on this. I'd also appreciate advice on how to print the integer types as something other than `_`.

rust-highfive commented:
r? @nrc

(rust_highfive has picked a reviewer for you, use r? to override)

bors commented:
:umbrella: The latest upstream changes (presumably #25091) made this pull request unmergeable. Please resolve the merge conflicts.

nrc commented:
I agree there is no need to start a 'note' block here.

As far as I know, you can't get a better representation of the int inference var than `_`, if the compiler can't infer the exact type, it is just a inference variable (well, type checking distinguishes between int vars and other inference vars, but error reporting doesn't).

This PR could do with some tests to make sure the errors don't regress.

(Also, needs a rebase)

nrc commented on src/librustc_typeck/check/_match.rs:
```diff
@@ -89,30 +89,46 @@ pub fn check_pat<'a, 'tcx>(pcx: &pat_ctxt<'a, 'tcx>,
             let rhs_ty = fcx.expr_ty(&**end);
 
-            let lhs_eq_rhs =
+            let compat = |t| ty::type_is_numeric(t) || ty::type_is_char(t);
```
numeric_or_char is a better name than compat

nrc commented on src/librustc_typeck/check/_match.rs:
```diff
@@ -89,30 +89,46 @@ pub fn check_pat<'a, 'tcx>(pcx: &pat_ctxt<'a, 'tcx>,
-            let numeric_or_char =
-                lhs_eq_rhs && (ty::type_is_numeric(lhs_ty) || ty::type_is_char(lhs_ty));
+            if !types_match {
+                return;
```
Should there be an error message in this case, or is it allowed?

michaelsproul replied:
`require_same_types` outputs a message, so just returning prevents duplication.

nrc replied:
Could you add a comment saying that please?

nrc commented on src/librustc_typeck/check/_match.rs:
```diff
@@ -89,30 +89,46 @@ pub fn check_pat<'a, 'tcx>(pcx: &pat_ctxt<'a, 'tcx>,
-                        span_err!(tcx.sess, begin.span, E0031,
-                            "mismatched types in range");
-                    }
+            match const_eval::compare_lit_exprs(tcx, &**begin, &**end, Some(lhs_ty)) {
```
don't need the `&**` on begin and end

nrc commented on src/librustc_typeck/check/_match.rs:
```diff
@@ -89,30 +89,46 @@ pub fn check_pat<'a, 'tcx>(pcx: &pat_ctxt<'a, 'tcx>,
-            } else {
-                span_err!(tcx.sess, begin.span, E0029,
-                          "only char and numeric types are allowed in range");
+                None => unreachable!()
```
Better to use span_bug than unreachable here

nrc commented on src/librustc_typeck/check/_match.rs:
```diff
@@ -89,30 +89,46 @@ pub fn check_pat<'a, 'tcx>(pcx: &pat_ctxt<'a, 'tcx>,
+                None => unreachable!()
             }
 
             fcx.write_ty(pat.id, lhs_ty);
```
I think you could write this type in the error cases too, it should let you type check further.

michaelsproul replied:
ok, right at the top is fine? just after `let lhs_ty = ...`?

nrc replied:
I think so, run the tests to make sure bad things don't happen

nrc commented on src/librustc_typeck/diagnostics.rs:
```diff
@@ -146,6 +146,46 @@ match d {
+"##,
+
+E0030: r##"
+When matching against a range the compiler verifies that the range is non-empty.
```
`When matching against a range, the compiler` (inserted comma)

michaelsproul commented:
I've addressed all your comments and pushed the commit, but I just noticed a failure in the compile-test results.

```
src/test/compile-fail/match-ill-type1.rs:13:9: 13:13 error: internal compiler error: error comparing literals in range pattern
src/test/compile-fail/match-ill-type1.rs:13         1...2_usize => 1, //~ ERROR mismatched types in range
```

Don't merge until I've fixed this...

michaelsproul commented on src/compiletest/runtest.rs:
```diff
@@ -954,14 +953,10 @@ fn check_expected_errors(expected_errors: Vec<errors::ExpectedError> ,
-        //                1234
-        // Should have 4 spaces: see issue 18946
-        line.starts_with("(")
+        line.starts_with(" ") || line.starts_with("(")
```
this change was required to support ` start type:` and ` end type:` (each with a preceding space).

michaelsproul commented on src/librustc_typeck/check/_match.rs:
```diff
@@ -83,42 +83,59 @@ pub fn check_pat<'a, 'tcx>(pcx: &pat_ctxt<'a, 'tcx>,
-                            "lower range bound must be less than upper");
-                    }
-                    None => {
-                        span_err!(tcx.sess, begin.span, E0031,
```
@nrc: I worked out that by removing this line I broke the match-ill-typed test I mentioned earlier. This error seems redundant as we've already run `require_same_types` by this stage. Is there a version that'll write a concrete type for the inferred int type, or fail? Or do you think I should just add this error back in?

nrc replied:
Do you get an error when you run that test? If we do get an error, just a different one, then you can just change the test and its fine (assuming the error is not less precise). If you get no error at all, I would be more concerned. I guess if this is causing inference to infer the type of the lhs of the range due to the explicit type on the rhs of the range (in the test), then it is Ok. If we're just sort of ignoring the problem, then it's not Ok.

michaelsproul replied:
I get an ICE because of the `span_bug` call I added in place of this.

The full example is this:

```rust
fn main() {
    match 1 {
        1 ... 2_usize => 1, //~ ERROR mismatched types in range
        _ => 2,
    };
}
```

I think the best behaviour would be to infer the type of the start argument to match the end.

michaelsproul replied:
I'll poke around in `infer` and see if I can get it to work :smile:

michaelsproul commented:
@nrc: I got it working! By computing a least-upper-bound for the begin and end types of the range and using it as a hint when evaluating and comparing the constants we get inference of the least-complete type from the most-complete one! As a bonus, this also removes the asymmetry whereby `lhs_ty` was used to type the whole pattern, because we can use the LUB instead. To reflect this change I moved the match-ill-typed example out of compile-fail and into run-pass, with a few added inference tests.

This is the output for the example in the PR description:

```
range.rs:4:9: 4:19 error: mismatched types in range:
 expected char,
    found integral variable [E0211]
range.rs:4         'a' ... 10 => (),
                   ^~~~~~~~~~
range.rs:5:9: 5:19 error: mismatched types in range:
 expected integral variable,
    found char [E0211]
range.rs:5         10 ... 'z' => (),
                   ^~~~~~~~~~
range.rs:6:9: 6:19 error: only char and numeric types are allowed in range patterns
 start type: &'static str
 end type: _ [E0029]
range.rs:6         "what" ... 10 => (),
                   ^~~~~~~~~~
range.rs:6:9: 6:19 help: run `rustc --explain E0029` to see a detailed explanation
range.rs:7:9: 7:26 error: only char and numeric types are allowed in range patterns
 start type: &'static str
 end type: &'static str [E0029]
range.rs:7         "what" ... "well" => (),
                   ^~~~~~~~~~~~~~~~~
range.rs:7:9: 7:26 help: run `rustc --explain E0029` to see a detailed explanation
range.rs:8:16: 8:25 error: only char and numeric types are allowed in range patterns
 start type: _
 end type: &'static str [E0029]
range.rs:8         10 ... "what" => ()
                          ^~~~~~~~~
range.rs:8:16: 8:25 help: run `rustc --explain E0029` to see a detailed explanation
error: aborting due to 5 previous errors
```

nrc commented on src/test/run-pass/match-range-infer.rs:
```diff
@@ -8,9 +8,17 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
```
Could you add a comment explaining what is being tested please?

nrc commented on src/librustc_typeck/check/_match.rs:
```diff
@@ -83,41 +83,67 @@ pub fn check_pat<'a, 'tcx>(pcx: &pat_ctxt<'a, 'tcx>,
+            // guarantee that the types are concrete and identical. We need to use the most
+            // specific type as a type hint.
+            // For example if we have 0 ... 5usize, we need to infer 0 : usize.
+            let type_hint = infer::common_supertype(
```
type_hint seems like a poor name - it is not a hint which we try to type check an expression with, it comes from type checking expressions without a hint. Maybe `common_type` or something?

michaelsproul replied:
Fixed.

nrc commented:
Other than the small nits, this looks good. I wanted to ask around a few more people that unifying the type here is OK (I'm pretty sure it is). I'll try and get this r+'ed tomorrow.

michaelsproul commented:
Ok great, cheers.

nikomatsakis commented on src/librustc_typeck/check/_match.rs:
```diff
@@ -83,41 +83,67 @@ pub fn check_pat<'a, 'tcx>(pcx: &pat_ctxt<'a, 'tcx>,
+            // guarantee that the types are concrete and identical. We need to use the most
+            // specific type as a type hint.
+            // For example if we have 0 ... 5usize, we need to infer 0 : usize.
+            let common_type = infer::common_supertype(
```
given that we've just unified the two types, computing a common supertype seems harmless but inefficient. If you just want to remove type variables from the type, `resolve_type_vars_if_possible` or some similar method seems preferable.

michaelsproul commented on src/librustc_typeck/check/_match.rs:
```diff
@@ -83,41 +83,64 @@ pub fn check_pat<'a, 'tcx>(pcx: &pat_ctxt<'a, 'tcx>,
-            fcx.write_ty(pat.id, lhs_ty);
+            // Now that we know the types can be unified we find the unified type and use
+            // it to type the entire expression.
+            let common_type = fcx.infcx().resolve_type_vars_if_possible(&lhs_ty);
```
@nikomatsakis, this works... there's no need to call it again with `rhs_ty` is there?

nikomatsakis replied:
@michaelsproul nope, that looks good to me. (calling it `rhs_ty` would yield the same result)

nrc commented:
@bors r+

bors commented:
:pushpin: Commit 25d0ef3 has been approved by `nrc`

<!-- @bors r=nrc 25d0ef347affe79557cdfbaa4e66144c7f476fc0 -->

bors commented:
:hourglass: Testing commit 25d0ef3 with merge 06c6b3c...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/5203), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/5212), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/5166), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/5192), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/5160), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/5233), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/5227), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/5199), [auto-win-gnu-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-nopt-t/builds/178), [auto-win-gnu-32-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/178), [auto-win-gnu-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-nopt-t/builds/178), [auto-win-gnu-64-opt](http://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/177)
